frontend/wasm/src/module/data_segments.rs

```diff
@@ -1,272 +1,380 @@
-            new_data.resize(padding as usize, 0);
-            new_data.extend_from_slice(&self.data);
-            self.data = new_data;
+    /// Pads the segment data to be aligned to WORD_SIZE_BYTES (16 bytes)
+    pub fn pad_to_word_alignment(&mut self) {
+        let remainder = self.data.len() % WORD_SIZE_BYTES;
+        if remainder != 0 {
+            let padding_size = WORD_SIZE_BYTES - remainder;
+            self.data.resize(self.data.len() + padding_size, 0);
         }
     }
+}
 
-    /// Check if this segment overlaps with another segment
-    pub fn overlaps_with(&self, other: &ResolvedDataSegment) -> bool {
-        let self_end = self.offset + self.data.len() as u32;
-        let other_end = other.offset + other.data.len() as u32;
+impl Debug for ResolvedDataSegment {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ResolvedDataSegment")
+            .field("offset", &self.offset)
+            .field("size", &self.data.len())
+            .field("name", &self.name)
+            .field("readonly", &self.readonly)
+            .field("data", &self.data.to_hex())
+            .finish()
+    }
+}
 
-        (self.offset < other_end) && (other.offset < self_end)
+/// Returns an error if any two segments overlap
+fn validate_no_overlaps(segments: &[ResolvedDataSegment]) -> WasmResult<()> {
+    assert!(segments.is_sorted_by_key(|s| s.offset), "Segments must be sorted by offset");
+    for window in segments.windows(2) {
+        let current = &window[0];
+        let next = &window[1];
+        if current.end_offset() > next.offset {
+            let error_msg = format!(
+                "Data segments overlap: segment '{}' at offset {:#x} (size {}) overlaps with \
+                 segment '{}' at offset {:#x}",
+                current.name,
+                current.offset,
+                current.data.len(),
+                next.name,
+                next.offset
+            );
+            return Err(Report::msg(WasmError::Unsupported(error_msg)));
+        }
     }
+    Ok(())
+}
 
-    /// Merge this segment with another segment, filling gap with zeros
-    pub fn merge_with(&mut self, other: ResolvedDataSegment) {
-        // Ensure this segment comes before the other
-        assert!(self.offset <= other.offset, "Segments must be sorted by offset");
+/// Merges segment metadata (names and readonly status)
+/// Returns a tuple with the merged name and whether all segments are readonly
+fn merge_metadata(segments: &[ResolvedDataSegment]) -> (String, bool) {
+    let mut merged_name = String::new();
+    let mut all_readonly = true;
 
-        let self_end = self.offset + self.data.len() as u32;
-        let gap = other.offset.saturating_sub(self_end);
+    for (i, segment) in segments.iter().enumerate() {
+        if i > 0 {
+            merged_name.push('_');
+        }
+        merged_name.push_str(&segment.name);
+        all_readonly = all_readonly && segment.readonly;
```

Comment:
We can't merge adjacent segments that don't have the same `readonly` status - we can only merge runs of segments with the same status.

I'm not sure how often we have multiple segments where some are readonly and others aren't, but we need to preserve the semantic distinction for sure. In the future, we'll need to enforce memory protection in some fashion, and we'll be reliant on the `readonly` flag to tell us whether that region of memory can be written to or not. We don't enforce it now, but if we don't preserve the information, it is useless.